### PR TITLE
feat: trainees list paginator

### PR DESCRIPTION
**Pull request by omarCreativeDev**
TISNEW-4186

- pagination feature for trainees list
- updated `IGetTraineesResponse` interface as per BE changes
- updated `getTrainees` FE mock endpoint
- fixed all outstanding warnings in unit testing around trainee details screen components
- removed redundant 2 tests from `AppComponent`

### CODE REVIEW

File: src/app/trainees/state/trainees.actions.ts [hunk 1]
```diff
@@ -26,3 +26,16 @@ export class SearchTrainees {
   static readonly type = "[Trainees] Search";
   constructor(public query: string) {}
 }
+
+export class PaginateTrainees {
+  static readonly type = "[Trainees] Page";
+  constructor(public pageIndex: number) {}
```
Comment (ojemuyiwa) [inline src/app/trainees/state/trainees.actions.ts:32]: just wondering: shouldn't there also be a, number of items per page? i.e. pageIndex and pageSize parameters? i.e. if we're defaulting to pageSize = 20

File: src/app/trainees/trainees.component.html [hunk 2]
```diff
@@ -5,3 +5,7 @@
 <div class="p-15">
   <router-outlet></router-outlet>
 </div>
+
```
Comment (ojemuyiwa) [inline src/app/trainees/trainees.component.html:8]: Tying component instances with routing may have an adverse effect with events. e.g. ngOnInit, etc will fire at separate instances out of sync if event based data. Just thinking aloud if clearFilters, pageChanges etc affect routerNagivationEnd, then data will be out of sync. If components are all loaded in tranee-list @inputs and outputs may keep them in sync? Maybe? trainee lists may keep subscribing to parameter changes (this.route.queryParamMap.subscribe || this.route.queryParams.subscribe) and fire actions appropriately? I will download this branch and see events sequences with sorting and page changing to see ... once again just thinking aloud

File: src/app/trainees/state/trainees.actions.ts (hunk 1 above)
Comment (omarCreativeDev) [reply src/app/trainees/state/trainees.actions.ts:32]: the `pageSize` of 20 is passed in via an input on the `<mat-paginator>`

Comment (ojemuyiwa) [reply src/app/trainees/state/trainees.actions.ts:32]: says pageSize=1 on <mat-paginator>

Comment (ojemuyiwa) [reply src/app/trainees/state/trainees.actions.ts:32]: And this is not passed to Uzairs code to return 20 items? Thats what I was alluding to

File: src/app/trainees/trainees.component.html (hunk 2 above)
Comment (omarCreativeDev) [reply src/app/trainees/trainees.component.html:8]: i dont think i quite understand. the route change is basically done to gracefully update the url which doesn't reinstantiate the component or fire ngOnInit either. 

the queryParams check via url snapshot is to cater for the scenario if `/trainees` route is viewed without any parameters it lets the store know how to update appropriately and subsequently make the api call with the correct default parameters

File: src/app/trainees/state/trainees.actions.ts (hunk 1 above)
Comment (omarCreativeDev) [reply src/app/trainees/state/trainees.actions.ts:32]: ah yea my bad, good spot. just updated the pr with 20

File: src/app/trainees/trainees.component.html (hunk 2 above)
Comment (ojemuyiwa) [reply src/app/trainees/trainees.component.html:8]: yeah i just checked, we're firing quite a bit in different events.

Screenshot 2020-04-07 at 11 35 57

Comment (ojemuyiwa) [reply src/app/trainees/trainees.component.html:8]: i'll pull-down and check again. I suspect we're talking about different things here

Comment (ojemuyiwa) [reply src/app/trainees/trainees.component.html:8]: > i dont think i quite understand. the route change is basically done to gracefully update the url which doesn't reinstantiate the component or fire ngOnInit either.
> 
> the queryParams check via url snapshot is to cater for the scenario if `/trainees` route is viewed without any parameters it lets the store know how to update appropriately and subsequently make the api call with the correct default parameters

In response: ngOnint always fires on route change if your component loads in router-outlet

Comment (omarCreativeDev) [reply src/app/trainees/trainees.component.html:8]: the BE work for pagination hasn't completed yet so the api response is sending back thousands of records and the ui therefore has a super long almost never ending scroll. i would imagine the paginator is out of view but not to worry as I've got another task to integrate the pagination with the BE once uzair is done. 

ngOnInit doesn't fire when navigating to the same route which is what is happening on this page

Comment (ojemuyiwa) [reply src/app/trainees/trainees.component.html:8]: Screenshot 2020-04-07 at 16 26 27

On Initial load, OnInit in pagination fires before OnInit in trainee-list.

This means is I hard code variables in url and hard refresh pagination control sends action first 
`if (params.pageIndex) {
      this.store.dispatch(new PaginateTrainees(params.pageIndex));
    } else {
      this.store.dispatch(new ResetTraineesPaginator());
    }`
then trainee-list sends its own OnInit
`if (params.active && params.direction) {
      this.store.dispatch(new SortTrainees(params.active, params.direction));
    } else {
      this.store.dispatch(new ResetTraineesSort());
    }

this.store.dispatch(new GetTrainees());
`

Comment (ojemuyiwa) [reply src/app/trainees/trainees.component.html:8]: Screenshot 2020-04-07 at 16 27 51

**Pull request merged**